codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```
@@ -0,0 +1,652 @@
+        writer.writeDocs(
+                String.format("%s defines the waiters for %s", clientName, waiterName)
+        );
+        writer.openBlock("type $L struct {", "}",
+                clientName, () -> {
+                    writer.write("");
+                    writer.write("client $L", InterfaceGenerator.getApiClientInterfaceName(operationSymbol));
+
+                    writer.write("");
+                    writer.write("options $L", generateWaiterOptionsName(waiterName));
+                });
+
+        writer.write("");
+
+        String constructorName = String.format("New%s", clientName);
+
+        Symbol waiterOptionsSymbol = SymbolUtils.createPointableSymbolBuilder(
+                generateWaiterOptionsName(waiterName)
+        ).build();
+
+        Symbol clientSymbol = SymbolUtils.createPointableSymbolBuilder(
+                clientName
+        ).build();
+
+        writer.writeDocs(
+                String.format("%s constructs a %s.", constructorName, clientName)
+        );
+        writer.openBlock("func $L(client $L, optFns ...func($P)) $P {", "}",
+                constructorName, InterfaceGenerator.getApiClientInterfaceName(operationSymbol),
+                waiterOptionsSymbol, clientSymbol, () -> {
+                    writer.write("options := $T{}", waiterOptionsSymbol);
+                    writer.addUseImports(SmithyGoDependency.TIME);
+
+                    // set defaults
+                    writer.write("options.MinDelay = $L * time.Second", waiter.getMinDelay());
+                    writer.write("options.MaxDelay = $L * time.Second", waiter.getMaxDelay());
+                    writer.write("options.Retryable = $L", generateRetryableName(waiterName));
+                    writer.write("");
+
+                    writer.openBlock("for _, fn := range optFns {",
+                            "}", () -> {
+                                writer.write("fn(&options)");
+                            });
+
```

Should we reset any of the defaults back if a user gives an invalid range? Like a negative value for min or max delay?

---

Since waiters are mostly sleeping, think it's better to error out if invalid delays override the default values. Switching back to default might cost them more than they wanted.   

Similar rationale behind why maxWaitTime is required now.  I can add validation to exit early if `maxdelay` and `mindelay` values are invalid. Thoughts?